ARCs/arc-0020.md

```diff
@@ -0,0 +1,747 @@
+          "name": "freeze_asset"
+        }
+      ],
+      "returns": {
+        "type": "bool"
+      }
+    },
+    {
+      "name": "get_account_is_frozen",
+      "readonly": true,
+      "args": [
+        {
+          "type": "asset",
+          "name": "freeze_asset"
+        },
+        {
+          "type": "account",
+          "name": "freeze_account"
+        }
+      ],
+      "returns": {
+        "type": "bool"
+      }
+    },
+    {
+      "name": "asset_destroy",
+      "args": [
+        {
+          "type": "asset",
+          "name": "destroy_asset"
+        }
+      ],
+      "returns": {
+        "type": "void"
+      }
+    },
+    {
+      "name": "get_circulating_supply",
+      "readonly": true,
+      "args": [
+        {
+          "type": "asset",
+          "name": "asset"
+        }
+      ],
+      "returns": {
+        "type": "uint64"
+      }
+    }
+  ]
+}
+```
+
+### Metadata
+
+#### ASA Metadata
+
+The ASA underlying a Smart ASA:
+
+- MUST be `DefaultFrozen`.
+- MUST specify the ID of the controlling Smart Contract (see below); and
+- MUST set the `ClawbackAddr` to the account of such Smart Contract.
+
+The metadata **MUST** be immutable.
+
+#### Specifying the controlling Smart Contract
+
+A Smart ASA MUST specify the ID of its controlling Smart Contract.
+
+If the Smart ASA also conforms to any ARC that supports additional `properties`
+([ARC-3](./arc-0003.md), [ARC-69](./arc-0069.md)), then it MUST include a
+`arc-20` key and set the corresponding value to a map, including the ID of the
+controlling Smart Contract as a value for the key `application-id`. For example:
+
+```javascript
+{
+  //...
+  "properties": {
+    //...
+    "arc-20": {
+      "application-id": 123
+    }
+  }
+  //...
+}
+```
+
+> To avoid ecosystem fragmentation this ARC does NOT propose any new method to
+> specify the metadata of an ASA. Instead, it only extends already
+> existing standards.
+
+### Handling opt in and close out
+
+A Smart ASA MUST require users to opt to the ASA and MAY require them to opt in
+to the controlling Smart Contract. This MAY be performed at two separate times.
+
+The reminder of this section is non-normative.
+
+> Smart ASAs SHOULD NOT require users to opt in to the controlling Smart
+> Contract, unless the implementation requires storing information into their
+> local schema (for instance, to implement [freezing](#asset-freeze); also see
+> [security considerations](#security-considerations)).
+>
+> Clients MAY inspect the local state schema of the controlling Smart Contract
+> to infer whether opt in is required.
+>
+> If a Smart ASA requires opt in, then clients SHOULD prevent users from closing
+> out the controlling Smart Contract unless they don't hold a balance for any of
+> the ASAs controlled by the Smart Contract.
+
+## Rationale
+
+This ARC builds on the strengths of the ASA to enable a Smart Contract to
+control its operations and flexibly re-configure its configuration.
+
```

Comment:
```suggestion

ARC-20 is meant to be an "extension" of ASAs, rather than a "replacement" of the existing ASA framework with an approach purely based on Smart Contract, _a-la Ethereum_. Regular ASAs should always be intended as _go-to_ solutions if they cover all the requirements of the specific use-case. The Smart ASA "extension" aims to piggyback on ASAs battle-tested security, their users and tooling adoption as well as preserving their backward compatibility with ASA-based ARCs (e.g. ARC-3, ARC-69, etc.).

```